src/runtime-rs/crates/resource/src/cgroups/mod.rs

```diff
@@ -35,30 +37,59 @@ pub struct CgroupArgs {
     pub config: TomlConfig,
 }
 
+#[derive(Default)]
 pub struct CgroupConfig {
     pub path: String,
     pub overhead_path: String,
     pub sandbox_cgroup_only: bool,
+    pub is_threaded_mode: bool,
+    pub specified_controllers: Option<Vec<String>>,
 }
 
 impl CgroupConfig {
     fn new(sid: &str, toml_config: &TomlConfig) -> Result<Self> {
-        let overhead_path = utils::gen_overhead_path(sid);
+        let mut overhead_path = utils::gen_overhead_path(sid);
         let spec = load_oci_spec()?;
-        let path = spec
+        let mut path = spec
             .linux
             // The trim of '/' is important, because cgroup_path is a relative path.
             .map(|linux| linux.cgroups_path.trim_start_matches('/').to_string())
             .unwrap_or_default();
 
+        path = utils::add_kata_prefix_for_cgroup_path(&path);
+        overhead_path = utils::add_kata_prefix_for_cgroup_path(&overhead_path);
+
+        // In cgroup v2, the sandbox and overhead cgroup in threaded mode are placed under
+        // the same cgroup in domain threaded mode.
+        // In this way, vCPU threads and VMM processes can be separated into two cgroups.
+        // For details, please refer to host cgroups design document
+        // https://github.com/kata-containers/kata-containers/blob/main/docs/design/host-cgroups.md.
+        let is_v2 = hierarchies::is_cgroup2_unified_mode();
+        let is_threaded_mode = is_v2 && !toml_config.runtime.sandbox_cgroup_only;
+        let specified_controllers = if is_threaded_mode {
+            Some(vec![String::from("cpuset"), String::from("cpu")])
+        } else {
+            None
+        };
+
+        let sandbox_path = if is_threaded_mode {
+            overhead_path = format!("{}/overhead", path);
+            format!("{}/sandbox", path)
+        } else {
+            path
+        };
```

Comment:
The code might be optimized as following: mut removal, reduction of assigned value.

```
// remove mut
let overhead_path = utils::gen_overhead_path(sid);
let spec = load_oci_spec()?;
// remove mut
let path = utils::add_kata_prefix_for_cgroup_path(&spec
// ...

// remove this
// path = utils::add_kata_prefix_for_cgroup_path(&path);
// remove this
// overhead_path = utils::add_kata_prefix_for_cgroup_path(&overhead_path);

// ...

let (path, overhead_path) = if is_threaded_mode {
    (format!("{}/sandbox", path), format!("{}/overhead", path))
} else {
    (path, utils::add_kata_prefix_for_cgroup_path(&utils::gen_overhead_path(sid)))
}
```

Reply:
I followed your suggestion, and wrapped two functions to process the path.